Patches relating to OpenMP imperfect loop collapse

jtb20 opened the pull request:
Various patches relating to internal issue SWDEV-229378 (upstream PR https://github.com/llvm/llvm-project/pull/96087)

jtb20 committed:
Use 'communicate' method for subprocess in LLNL test script

This patch uses the subprocess 'communicate' method in the LLNL Python
test script, as recommended in the documentation, and adds some
return-code checks.  This was mostly done as a side-effect of
investigating another problem, but is probably helpful anyway.

jtb20 committed:
Add timeout detection, and also increase timeout for LLNL tests

This patch fixes a mysterious bug with the LLNL test script: if it takes
more than two minutes to run, it silently exits without producing any
output (with logging enabled).  This can happen for debug builds of the
compiler, even on a fast machine.

This happens because the underlying Python script buffers output because
it knows it is writing to a pipe.  But if the 'timeout' command kills
that script before it has completed, all its output is then discarded.

The fix is twofold: firstly the "timed-out" condition is now detected
in the check_LLNL.sh script, and secondly the timeout time is increased.

jtb20 committed:
XFAIL support for LLNL/openmp5.0-tests tests

The test test_imperfect_loop.cxx contains a loop collapse operation that
cannot be executed safely in parallel, so should be marked as expected
to fail, somehow.  This patch adds a basic way to do that.

jtb20 committed:
Split imperfect-loop-collapse into usm/non-usm variants

This patch splits the smoke-fails/imperfect-loop-collapse test into two
variants, one of which does and one which doesn't need unified shared
memory (which is orthogonal to the main purpose of the test).  It also
fixes the HSA_XNACK setting for smoke-fails/imperfect-loop-collapse-usm
version, so both versions now work.

estewart08 commented on bin/rocm-test/scripts/parse_LLNL.sh:
```diff
@@ -49,6 +52,10 @@ function parse(){
       fi
       [[ "$line" =~ $testname_regex ]]
       llnltest=${BASH_REMATCH[1]}
+      if [ "$outfile" = "failing-tests.txt" ] && \
```
Need the rest of the conditional operator here. I think you are going for `==`.

jtb20 replied:
I think string equality with "test" is done with =, not ==. "man test" says:

```
       STRING1 = STRING2
              the strings are equal
```
test is [ of course. This is two conditions, a && b, where a is the [...] part and b is the grep on the following line -- hence the \ continuation.

estewart08 replied:
Looks like both `=` and `==` are accepted by bash but `=` is the standard for test. LGTM, thanks.

estewart08 commented on bin/rocm-test/scripts/parse_LLNL.sh:
```diff
@@ -49,6 +52,10 @@ function parse(){
       [[ "$line" =~ $testname_regex ]]
       llnltest=${BASH_REMATCH[1]}
+      if [ "$outfile" = "failing-tests.txt" ] && \
+	 grep -Fq "$llnltest" xfail-list then
```
Missing a semi-colon here after `xfail-list`.

jtb20 replied:
...huh, but yeah. I wonder why it didn't complain! I'll fix that, thanks.

estewart08 commented:
Just to confirm, we expect `openmp5.0-tests/test_imperfect_loop_warning.cxx` to pass once the llvm patch lands?

jtb20 commented:
> Just to confirm, we expect `openmp5.0-tests/test_imperfect_loop_warning.cxx` to pass once the llvm patch lands?

Yes (although the "test_imperfect_loop" of course xfails, in this hackish new way).

jtb20 committed:
Robustify parse_LLNL.sh script a bit

This patch adds some error checking to the parse_LLNL.sh script, and
avoids parsing `ls` output when looking for the results file.

jtb20 commented:
Thank you for the approval!

I can't merge this though -- it _could_ wait until the upstream imperfect loop analysis option lands on amd-staging, but it probably doesn't need to (just the "remark" output-scanning test will fail for now, I guess).
